Add initial wiki page for GSoC 2024. (#246)
oppia/oppia-web-developer-docs: Google-Summer-of-Code-2024.md

  - [Tips for writing a good project plan](#tips-for-writing-a-good-project-plan)
  - [What should applicants expect from mentors in a proposal review?](#what-should-applicants-expect-from-mentors-in-a-proposal-review)
  - [Sample proposals from past years](#sample-proposals-from-past-years)
- [Selection Criteria](#selection-criteria)
- [Communication](#communication)
- [Oppia's Project Ideas List](#oppias-project-ideas-list)

Oppia is planning to participate in [Google Summer of Code 2024](https://summerofcode.withgoogle.com/)! GSoC is a global program which offers an opportunity to discover and work with open source organizations to post-secondary students and other non-experienced contributors. The contributions are supported by a stipend. Contributors work closely with one or more mentors from an open source organization in order to implement either a project idea by the organization, or a proposal of their own.

In order to receive updates about GSoC at Oppia, please subscribe to the [Oppia GSoC Announce](https://groups.google.com/g/oppia-gsoc-announce) mailing list and keep an eye on [GitHub Discussions](https://github.com/oppia/oppia/discussions).

> This year, Oppia will follow the standard GSoC timeline: projects will have 6 weeks for each milestone, with a PM demo and internal evaluation after 5 weeks. Please refer to the [Dates and Deadlines](#dates-and-deadlines) section below for more details.

Reviewer: Maybe add the full form of PM here!
```suggestion
This year, Oppia will follow the standard GSoC timeline: projects will have 6 weeks for each milestone, with a Product Manager (PM) demo and internal evaluation after 5 weeks. Please refer to the [Dates and Deadlines](#dates-and-deadlines) section below for more details.
```

Author: Done.

Also, please note that acceptance into GSoC isn't a prerequisite for becoming an Oppia contributor. The Oppia project is run by the community for the community, and we warmly welcome anyone who'd like to help out! You can get started by following the instructions [here](https://github.com/oppia/oppia/wiki).

## Contributors

GSoC is an excellent opportunity for new contributors to get paid to work on an open source project. If you're interested in applying as a contributor, you should definitely read the following resources:

-   [Google Summer of Code contributor guide](https://google.github.io/gsocguides/student/)
-   [Google's list of resources](https://developers.google.com/open-source/gsoc/resources/)
-   [GSoC FAQ](https://developers.google.com/open-source/gsoc/faq)

## Getting started

Welcome! If you're interested in applying to work with Oppia for GSoC, please follow these steps:

> 1. Sign up to the [oppia-gsoc-announce@](https://groups.google.com/forum/#!forum/oppia-gsoc-announce) mailing list in order to receive important notifications about Oppia's participation in GSoC. Also, turn on notifications for [Github Discussions](https://github.com/oppia/oppia/discussions) since announcements may be made there as well. Make sure to set your preferences for both of these correctly so that you actually get the emails!

Reviewer: ```suggestion
1. Sign up to the [oppia-gsoc-announce@](https://groups.google.com/forum/#!forum/oppia-gsoc-announce) mailing list in order to receive important notifications about Oppia's participation in GSoC. Also, turn on notifications for [GitHub Discussions](https://github.com/oppia/oppia/discussions) since announcements may be made there as well. Make sure to set your preferences for both of these correctly so that you actually get the emails!
```

Author: I dropped this since we are doing announcements only through the mailing list.

2. Get a better understanding of what Oppia is about:
    - Read the [user documentation](http://oppia.github.io/#/) to become familiar with important concepts like explorations and interactions.
    - Play some lessons on [Oppia.org](https://www.oppia.org/learn/math), which hosts a live instance of Oppia.

3. To get started with development, read and follow the instructions in the contributors' guide carefully ([Oppia Web](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#setting-things-up), [Oppia Android](https://github.com/oppia/oppia-android/wiki/Contributing-to-Oppia-android)).

> 4. Do one or more starter projects to become familiar with the contribution process. This will help us get an idea of what it's like to work with you. It will also help you get a better understanding of the codebase and our development process, which may help with writing a good project proposal. Once you've merged at least 2 pull requests, you will get an invitation to become a collaborator to the Oppia repository and be officially onboarded! **This step is a prerequisite to applying for GSoC.**

Reviewer: Do we always onboard contributors with 2 PRs (irrespective of the strength of the PR)? [Not sure what we follow these days]

Author: Yes, that is correct (though onboarding mentors have discretion to ask for an additional PR if the first two are trivial).

  - How to manage my time well, and how to achieve deadlines.
  - How to properly plan a project before implementing it.
  - How to give, respond to and understand reviews.
  - How to effectively convey ideas.
  - How to write a good project proposal.
- Communication and personal development
  - How to seek help when needed and overcome challenges.
  - How to reach out to people, work with them, and help solve each other's problems.
  - How to get myself unblocked.
  - Putting forward my thoughts more systematically so that others can understand me well.
  - How to work with a large community which is spread over different time zones.
  - I feel more confident while joining online meetings.
  - I have become a better developer, not only in term of technical skills but in thinking of actual application of the built product and the edge case scenarios that the user might face.

Contributors have also told us why they continue to stay engaged with the project after GSoC ends:

- Community
  - I had a great experience with Oppia and would like to continue contributing to the project. I'd like to help newer contributors too, (for example by reviewing their code when I get the chance).

>   - The organisation is active and have strong community bond.

Reviewer: ```suggestion
  - The organisation is active and has a strong community bond.
```

Author: Done.
